File: crates/harness-server/src/automation_routes.rs
use axum::{
    extract::{Path, State},
    response::{IntoResponse, Response},
    routing::{get, post},
    Json, Router,
};
use harness_automation::{
    parse_rfc3339, AutomationStatus, AutomationTask, NewAutomationTask, ScheduleSpec,
};
use serde::{Deserialize, Serialize};
use serde_json::json;

use crate::automation_runtime::{is_running, spawn_automation_run, RunTrigger};
use crate::state::AppState;

pub(crate) fn router() -> Router<AppState> {
    Router::new()
        .route(
            "/v1/automations",
            get(list_automations).post(create_automation),
        )
        .route(
            "/v1/automations/:id",
            get(get_automation)
                .patch(patch_automation)
                .delete(delete_automation),
        )
        .route("/v1/automations/:id/run", post(run_automation_now))
}

#[derive(Debug, Serialize)]
struct AutomationsResponse {
    items: Vec<AutomationTask>,
}

#[derive(Debug, Deserialize)]
struct CreateAutomationRequest {
    title: String,
    prompt: String,
    schedule: ScheduleSpec,
    #[serde(default)]
    status: Option<AutomationStatus>,
    #[serde(default)]
    provider: Option<String>,
    #[serde(default)]
    model: Option<String>,
    #[serde(default)]
    conversation_id: Option<String>,
}

#[derive(Debug, Default, Deserialize)]
struct PatchAutomationRequest {
    #[serde(default)]
    title: Option<String>,
    #[serde(default)]
    prompt: Option<String>,
    #[serde(default)]
    schedule: Option<ScheduleSpec>,
    #[serde(default)]
    status: Option<AutomationStatus>,
    #[serde(default)]
    provider: Option<Option<String>>,
    #[serde(default)]
    model: Option<Option<String>>,
    #[serde(default)]
    conversation_id: Option<Option<String>>,
}

async fn list_automations(State(state): State<AppState>) -> Response {
    let Some(store) = state.automations.as_ref() else {
        return unavailable();
    };
    match store.list().await {
        Ok(items) => Json(AutomationsResponse { items }).into_response(),
        Err(error) => server_error(error.to_string()),
    }
}

async fn create_automation(
    State(state): State<AppState>,
    Json(req): Json<CreateAutomationRequest>,
) -> Response {
    let Some(store) = state.automations.as_ref() else {
        return unavailable();
    };
    if let Err(error) = validate_schedule(&req.schedule) {
        return bad_request(error);
    }
    if req.title.trim().is_empty() || req.prompt.trim().is_empty() {
        return bad_request("title and prompt are required");
    }
    let task = AutomationTask::new(NewAutomationTask {
        title: req.title.trim().to_string(),
        prompt: req.prompt,
        schedule: req.schedule,
        status: req.status,
        provider: clean_opt(req.provider),
        model: clean_opt(req.model),
        conversation_id: clean_opt(req.conversation_id),
    });
    match store.upsert(&task).await {
        Ok(()) => (axum::http::StatusCode::CREATED, Json(task)).into_response(),
        Err(error) => server_error(error.to_string()),
    }
}

async fn get_automation(State(state): State<AppState>, Path(id): Path<String>) -> Response {
    let Some(store) = state.automations.as_ref() else {
        return unavailable();
    };
    match store.get(&id).await {
        Ok(Some(task)) => Json(task).into_response(),
        Ok(None) => not_found(),
        Err(error) => server_error(error.to_string()),
    }
}

async fn patch_automation(
    State(state): State<AppState>,
    Path(id): Path<String>,
    Json(req): Json<PatchAutomationRequest>,
) -> Response {
    let Some(store) = state.automations.as_ref() else {
        return unavailable();
    };
    let mut task = match store.get(&id).await {
        Ok(Some(task)) => task,
        Ok(None) => return not_found(),
        Err(error) => return server_error(error.to_string()),
    };
    if let Some(title) = req.title {
        if title.trim().is_empty() {
            return bad_request("title cannot be empty");
        }
        task.title = title.trim().to_string();
    }
    if let Some(prompt) = req.prompt {
        if prompt.trim().is_empty() {
            return bad_request("prompt cannot be empty");
        }
        task.prompt = prompt;
    }
    if let Some(schedule) = req.schedule {
        if let Err(error) = validate_schedule(&schedule) {
            return bad_request(error);
        }
        task.schedule = schedule;
        if !is_running(&task) {
            task.recompute_next_run(chrono::Utc::now());
        }
    }
    if let Some(status) = req.status {
        task.status = status;
    }
    if let Some(provider) = req.provider {
        task.provider = clean_opt(provider);
    }
    if let Some(model) = req.model {
        task.model = clean_opt(model);
    }
    if let Some(conversation_id) = req.conversation_id {
        task.conversation_id = clean_opt(conversation_id);
    }
    task.updated_at = harness_automation::to_rfc3339(chrono::Utc::now());
    match store.upsert(&task).await {
        Ok(()) => Json(task).into_response(),
        Err(error) => server_error(error.to_string()),
    }
}

async fn delete_automation(State(state): State<AppState>, Path(id): Path<String>) -> Response {
    let Some(store) = state.automations.as_ref() else {
        return unavailable();
    };
    match store.delete(&id).await {
        Ok(true) => axum::http::StatusCode::NO_CONTENT.into_response(),
        Ok(false) => not_found(),
        Err(error) => server_error(error.to_string()),
    }
}

async fn run_automation_now(State(state): State<AppState>, Path(id): Path<String>) -> Response {
    let Some(store) = state.automations.as_ref() else {
        return unavailable();
    };
    let task = match store.get(&id).await {
        Ok(Some(task)) => task,
        Ok(None) => return not_found(),
        Err(error) => return server_error(error.to_string()),
    };
    if is_running(&task) {
        return (
            axum::http::StatusCode::CONFLICT,
            Json(json!({ "error": "automation is already running" })),
        )
            .into_response();
    }
    // Reserve synchronously so a manual trigger can't race a just-spawned
    // scheduled run (whose persisted `Running` flag may not have landed
    // yet). A failed claim means a run is already in flight.
    let Some(claim) = state.automation_claims.try_claim(&task.id) else {
        return (
            axum::http::StatusCode::CONFLICT,
            Json(json!({ "error": "automation is already running" })),
        )
            .into_response();
    };
    spawn_automation_run(state, task.clone(), RunTrigger::Manual, claim);
    (axum::http::StatusCode::ACCEPTED, Json(task)).into_response()
}

fn validate_schedule(schedule: &ScheduleSpec) -> Result<(), &'static str> {
    match schedule {
        ScheduleSpec::Once { run_at } => parse_rfc3339(run_at)
            .map(|_| ())
            .ok_or("schedule.once.run_at must be RFC3339"),
        ScheduleSpec::Interval {
            every_seconds,
            start_at,
        } => {
            if *every_seconds == 0 {
                return Err("schedule.interval.every_seconds must be greater than zero");
            }
            // Cap the interval at ~10 years. Beyond a sane upper bound the value is
            // almost certainly a mistake, and very large intervals push the next-run
            // arithmetic toward the edge of chrono's representable range.
            const MAX_EVERY_SECONDS: u64 = 60 * 60 * 24 * 366 * 10;
            if *every_seconds > MAX_EVERY_SECONDS {
                return Err("schedule.interval.every_seconds is too large (max ~10 years)");
            }
            if start_at
                .as_deref()
                .is_some_and(|value| parse_rfc3339(value).is_none())
            {
                return Err("schedule.interval.start_at must be RFC3339");
            }
            Ok(())
        }
    }
}

fn clean_opt(value: Option<String>) -> Option<String> {
    value.and_then(|s| {
        let trimmed = s.trim();
        (!trimmed.is_empty()).then(|| trimmed.to_string())
    })
}

fn unavailable() -> Response {
    (
        axum::http::StatusCode::SERVICE_UNAVAILABLE,
        Json(json!({ "error": "automation store not configured" })),
    )
        .into_response()
}

fn bad_request(message: impl Into<String>) -> Response {
    (
        axum::http::StatusCode::BAD_REQUEST,
        Json(json!({ "error": message.into() })),
    )
        .into_response()
}

fn not_found() -> Response {
    (
        axum::http::StatusCode::NOT_FOUND,
        Json(json!({ "error": "automation not found" })),
    )
        .into_response()
}

fn server_error(message: String) -> Response {
    (
        axum::http::StatusCode::INTERNAL_SERVER_ERROR,
        Json(json!({ "error": message })),
    )
        .into_response()
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::Arc;

    struct StubLlm;

    #[async_trait::async_trait]
    impl harness_core::LlmProvider for StubLlm {
        async fn complete(
            &self,
            _: harness_core::ChatRequest,
        ) -> Result<harness_core::ChatResponse, harness_core::Error> {
            Err(harness_core::Error::Provider("stub".into()))
        }
    }

    fn mk_state() -> AppState {
        let cfg = harness_core::AgentConfig::new("stub-model");
        let agent = Arc::new(harness_core::Agent::new(Arc::new(StubLlm) as _, cfg));
        AppState::new(agent)
    }

    async fn read_json(resp: Response) -> serde_json::Value {
        let bytes = axum::body::to_bytes(resp.into_body(), 1024 * 1024)
            .await
            .unwrap();
        serde_json::from_slice(&bytes).unwrap()
    }

    #[tokio::test]
    async fn list_returns_503_when_store_absent() {
        let resp = list_automations(State(mk_state())).await;
        assert_eq!(resp.status(), axum::http::StatusCode::SERVICE_UNAVAILABLE);
    }

    #[tokio::test]
    async fn create_then_list_round_trip() {
        let state =
            mk_state().with_automation_store(Arc::new(harness_store::MemoryAutomationStore::new()));
        let create = create_automation(
            State(state.clone()),
            Json(CreateAutomationRequest {
                title: "Daily digest".into(),
                prompt: "Summarise the project".into(),
                schedule: ScheduleSpec::Interval {
                    every_seconds: 3600,
                    start_at: None,
                },
                status: None,
                provider: None,
                model: None,
                conversation_id: None,
            }),
        )
        .await;
        assert_eq!(create.status(), axum::http::StatusCode::CREATED);
        let created = read_json(create).await;
        assert_eq!(created["title"], "Daily digest");

        let list = list_automations(State(state)).await;
        let body = read_json(list).await;
        let items = body["items"].as_array().unwrap();
        assert_eq!(items.len(), 1);
        assert_eq!(items[0]["prompt"], "Summarise the project");
    }

    #[tokio::test]
    async fn create_rejects_invalid_schedule() {
        let state =
            mk_state().with_automation_store(Arc::new(harness_store::MemoryAutomationStore::new()));
        let resp = create_automation(
            State(state),
            Json(CreateAutomationRequest {
                title: "Bad".into(),
                prompt: "Run".into(),
                schedule: ScheduleSpec::Once {
                    run_at: "not-a-date".into(),
                },
                status: None,
                provider: None,
                model: None,
                conversation_id: None,
            }),
        )
        .await;
        assert_eq!(resp.status(), axum::http::StatusCode::BAD_REQUEST);
    }

    #[tokio::test]
    async fn create_rejects_oversized_interval() {
        let state =
            mk_state().with_automation_store(Arc::new(harness_store::MemoryAutomationStore::new()));
        let resp = create_automation(
            State(state),
            Json(CreateAutomationRequest {
                title: "Huge".into(),
                prompt: "Run".into(),
                schedule: ScheduleSpec::Interval {
                    every_seconds: u64::MAX,
                    start_at: None,
                },
                status: None,
                provider: None,
                model: None,
                conversation_id: None,
            }),
        )
        .await;
        assert_eq!(resp.status(), axum::http::StatusCode::BAD_REQUEST);
    }
}

File: packages/server/src/automation-routes.ts
// REST surface for scheduled automations. Ported from
// harness-server/src/automation_routes.rs.
//
// Every handler 503s when `state.automations` is unset, so tests / specialised
// binaries that don't wire the store get a clean "not configured" signal
// instead of a panic — mirrors the conversations-routes `requireStore` pattern.
//
//   - GET    /v1/automations          — list every task (newest-updated first).
//   - POST   /v1/automations          — create (validates schedule + title/prompt) → 201.
//   - GET    /v1/automations/:id       — fetch one (404 when absent).
//   - PATCH  /v1/automations/:id       — partial update (recomputes next_run when
//                                        the schedule changes and the task isn't running).
//   - DELETE /v1/automations/:id       — hard delete → 204 / 404.
//   - POST   /v1/automations/:id/run   — on-demand trigger → 202 / 404 / 409.
//
// The on-demand trigger reserves a process-wide {@link AutomationClaims} slot
// synchronously before spawning, so a manual trigger can't race a just-spawned
// scheduled run (the persisted `running` flag may not have landed yet) — exactly
// the ordering the Rust handler uses (`is_running` 409 → `try_claim` 409 → spawn).
//
// Node adaptation: the Rust handler reads `state.automation_claims` +
// `state.build_agent(provider, model)`. The Node `AppState` carries neither and
// must not be edited from this step, so the runtime owns a process-wide
// {@link sharedAutomationClaims} singleton (shared with the scheduler loop) and
// the route builds an {@link AutomationDeps} from `state.createAgent` for the
// trigger. Per-task provider/model overrides aren't honoured by the current
// `createAgent` signature; the run still proceeds on the default agent.
import type { FastifyInstance, FastifyReply } from "fastify";
import { errorText, type Agent } from "@jarvis/core";
import {
  newAutomationTask,
  parseTimestamp,
  recomputeNextRun,
  toRfc3339,
  type AutomationStatus,
  type AutomationStore,
  type AutomationTask,
  type ScheduleSpec,
} from "@jarvis/automation";
import type { AppState } from "./state.ts";
import {
  isRunning,
  triggerAutomationNow,
  type AutomationDeps,
} from "./automation-runtime.ts";

// ---------- request bodies --------------------------------------------------

interface CreateAutomationRequest {
  title?: unknown;
  prompt?: unknown;
  schedule?: unknown;
  status?: unknown;
  provider?: unknown;
  model?: unknown;
  conversation_id?: unknown;
}

interface PatchAutomationRequest {
  title?: unknown;
  prompt?: unknown;
  schedule?: unknown;
  status?: unknown;
  // `provider` / `model` / `conversation_id` use double-Option in Rust: present
  // with a string sets it, present with null clears it, absent leaves it as-is.
  provider?: unknown;
  model?: unknown;
  conversation_id?: unknown;
}

// ---------- store guard -----------------------------------------------------

/** Return the automation store, or send a 503 and return undefined. */
function requireStore(state: AppState, reply: FastifyReply): AutomationStore | undefined {
  if (!state.automations) {
    reply.code(503).send({ error: "automation store not configured" });
    return undefined;
  }
  return state.automations;
}

// ---------- helpers ---------------------------------------------------------

/**
 * Upper bound for `interval.every_seconds` — 100 years. Longer than any real
 * schedule, and small enough that `now + every_seconds*1000` stays well within
 * the valid epoch-ms range (max ≈ 8.64e15) so {@link toRfc3339} never throws.
 */
const MAX_INTERVAL_SECONDS = 100 * 365 * 24 * 60 * 60;

/**
 * Validate a {@link ScheduleSpec}. Returns an error string when invalid,
 * `undefined` when valid. Mirrors Rust's `validate_schedule`.
 */
function validateSchedule(schedule: unknown): string | undefined {
  if (typeof schedule !== "object" || schedule === null) {
    return "schedule is required";
  }
  if ("once" in schedule) {
    const once = (schedule as { once?: unknown }).once;
    const runAt = (once as { run_at?: unknown } | undefined)?.run_at;
    if (typeof runAt !== "string" || parseTimestamp(runAt) === undefined) {
      return "schedule.once.run_at must be RFC3339";
    }
    return undefined;
  }
  if ("interval" in schedule) {
    const interval = (schedule as { interval?: unknown }).interval as
      | { every_seconds?: unknown; start_at?: unknown }
      | undefined;
    const every = interval?.every_seconds;
    if (typeof every !== "number" || !Number.isFinite(every) || every <= 0) {
      return "schedule.interval.every_seconds must be greater than zero";
    }
    // Upper bound: `scheduleNextAfter` computes `now + every_seconds*1000`, and a
    // large-but-finite value (e.g. 9e18) overflows the valid epoch-ms range so
    // `toRfc3339(new Date(nextMs))` throws `RangeError`. Cap well below overflow
    // (100 years is longer than any real interval) so the create/update handler
    // returns a clean 400 instead of a 500, and the scheduler tick can never
    // throw on a persisted task.
    if (every > MAX_INTERVAL_SECONDS) {
      return `schedule.interval.every_seconds must be <= ${MAX_INTERVAL_SECONDS}`;
    }
    const startAt = interval?.start_at;
    if (startAt !== undefined && startAt !== null) {
      if (typeof startAt !== "string" || parseTimestamp(startAt) === undefined) {
        return "schedule.interval.start_at must be RFC3339";
      }
    }
    return undefined;
  }
  return "schedule must be `once` or `interval`";
}

/**
 * Trim a string, returning undefined when it's empty/whitespace-only or not a
 * string. Mirrors Rust's `clean_opt`.
 */
function cleanOpt(value: unknown): string | undefined {
  if (typeof value !== "string") return undefined;
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

/** A non-empty trimmed string, else undefined. */
function nonEmptyTrimmed(value: unknown): string | undefined {
  if (typeof value !== "string") return undefined;
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

function isAutomationStatus(value: unknown): value is AutomationStatus {
  return value === "enabled" || value === "paused";
}

/** Current time as the canonical RFC-3339 millis+`Z` wire form. */
function nowStamp(): string {
  return toRfc3339(Date.now());
}

/**
 * Build the runtime deps for the on-demand trigger from `state`. `createAgent`
 * doesn't yet take provider/model overrides, so they're dropped here (the run
 * proceeds on the default agent) — a forward-compat seam matching the workflow
 * runtime's per-step-model degradation.
 */
function triggerDeps(state: AppState, store: AutomationStore): AutomationDeps {
  return {
    automations: store,
    store: state.store,
    buildAgent: (): Agent => state.createAgent(),
  };
}

// ---------- routes ----------------------------------------------------------

export function registerAutomationRoutes(app: FastifyInstance, state: AppState): void {
  // list
  app.get("/v1/automations", async (_req, reply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    try {
      const items = await store.list();
      return reply.send({ items });
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
  });

  // create
  app.post("/v1/automations", async (req, reply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    const body = (req.body ?? {}) as CreateAutomationRequest;

    const scheduleError = validateSchedule(body.schedule);
    if (scheduleError !== undefined) {
      return reply.code(400).send({ error: scheduleError });
    }
    const title = typeof body.title === "string" ? body.title.trim() : "";
    const prompt = typeof body.prompt === "string" ? body.prompt : "";
    if (title.length === 0 || prompt.trim().length === 0) {
      return reply.code(400).send({ error: "title and prompt are required" });
    }
    if (body.status !== undefined && !isAutomationStatus(body.status)) {
      return reply.code(400).send({ error: "status must be `enabled` or `paused`" });
    }

    const input = {
      title,
      prompt,
      schedule: body.schedule as ScheduleSpec,
      ...(isAutomationStatus(body.status) ? { status: body.status } : {}),
      ...(cleanOpt(body.provider) !== undefined ? { provider: cleanOpt(body.provider)! } : {}),
      ...(cleanOpt(body.model) !== undefined ? { model: cleanOpt(body.model)! } : {}),
      ...(cleanOpt(body.conversation_id) !== undefined
        ? { conversation_id: cleanOpt(body.conversation_id)! }
        : {}),
    };
    const task = newAutomationTask(input);
    try {
      await store.upsert(task);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
    return reply.code(201).send(task);
  });

  // get one
  app.get("/v1/automations/:id", async (req, reply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    const id = (req.params as { id: string }).id;
    try {
      const task = await store.get(id);
      if (!task) return reply.code(404).send({ error: "automation not found" });
      return reply.send(task);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
  });

  // patch
  app.patch("/v1/automations/:id", async (req, reply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    const id = (req.params as { id: string }).id;
    let task: AutomationTask | undefined;
    try {
      task = await store.get(id);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
    if (!task) return reply.code(404).send({ error: "automation not found" });

    const body = (req.body ?? {}) as PatchAutomationRequest;

    if (body.title !== undefined) {
      const title = nonEmptyTrimmed(body.title);
      if (title === undefined) {
        return reply.code(400).send({ error: "title cannot be empty" });
      }
      task.title = title;
    }
    if (body.prompt !== undefined) {
      if (typeof body.prompt !== "string" || body.prompt.trim().length === 0) {
        return reply.code(400).send({ error: "prompt cannot be empty" });
      }
      task.prompt = body.prompt;
    }
    if (body.schedule !== undefined) {
      const scheduleError = validateSchedule(body.schedule);
      if (scheduleError !== undefined) {
        return reply.code(400).send({ error: scheduleError });
      }
      task.schedule = body.schedule as ScheduleSpec;
      // Only re-anchor next_run when the task isn't mid-run — a running task's
      // next_run was already advanced when it was picked up.
      if (!isRunning(task)) {
        recomputeNextRun(task, nowStamp());
      }
    }
    if (body.status !== undefined) {
      if (!isAutomationStatus(body.status)) {
        return reply.code(400).send({ error: "status must be `enabled` or `paused`" });
      }
      task.status = body.status;
    }
    // provider / model / conversation_id: present clears-or-sets, absent leaves.
    if (body.provider !== undefined) {
      const v = cleanOpt(body.provider);
      if (v === undefined) delete task.provider;
      else task.provider = v;
    }
    if (body.model !== undefined) {
      const v = cleanOpt(body.model);
      if (v === undefined) delete task.model;
      else task.model = v;
    }
    if (body.conversation_id !== undefined) {
      const v = cleanOpt(body.conversation_id);
      if (v === undefined) delete task.conversation_id;
      else task.conversation_id = v;
    }
    task.updated_at = nowStamp();
    try {
      await store.upsert(task);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
    return reply.send(task);
  });

  // delete
  app.delete("/v1/automations/:id", async (req, reply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    const id = (req.params as { id: string }).id;
    try {
      const deleted = await store.delete(id);
      if (!deleted) return reply.code(404).send({ error: "automation not found" });
      return reply.code(204).send();
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
  });

  // on-demand trigger
  app.post("/v1/automations/:id/run", async (req, reply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    const id = (req.params as { id: string }).id;
    const outcome = await triggerAutomationNow(triggerDeps(state, store), id);
    switch (outcome.kind) {
      case "accepted":
        return reply.code(202).send(outcome.task);
      case "not_found":
        return reply.code(404).send({ error: "automation not found" });
      case "conflict":
        return reply.code(409).send({ error: "automation is already running" });
      case "error":
        return reply.code(500).send({ error: outcome.error });
    }
  });
}
